src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ConnectionHandler.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Transport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal
+{
+    public class ConnectionHandler<TContext> : IConnectionHandler, IDisposable
+    {
+        // Base32 encoding - in ascii sort order for easy text based sorting
+        private static readonly string _encode32Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUV";
+
+        // Seed the _lastConnectionId for this application instance with
+        // the number of 100-nanosecond intervals that have elapsed since 12:00:00 midnight, January 1, 0001
+        // for a roughly increasing _requestId over restarts
+        private static long _lastConnectionId = DateTime.UtcNow.Ticks;
+
+        private readonly ServiceContext _serviceContext;
+        private readonly IHttpApplication<TContext> _application;
+        private readonly PipeFactory _pipeFactory;
+
+        public ConnectionHandler(ServiceContext serviceContext, IHttpApplication<TContext> application)
+        {
+            _serviceContext = serviceContext;
+            _application = application;
+            _pipeFactory = new PipeFactory();
```

Comment:
This needs to be passed into OnConnection. The transport may want a pool per thread.